libvast/src/option_declaration_set.cpp

```diff
@@ -0,0 +1,296 @@
+    [&](const auto& arg) {
+      using arg_type = std::decay_t<decltype(arg)>;
+      auto x = to<arg_type>(value);
+      if (!x)
+        // FIXME: We lose valuable error informaton here.
+        return std::make_pair(parse_state::failed_to_parse_argument,
+                              default_value());
+      return std::make_pair(parse_state::successful, data{*x});
+    },
+    // TODO: These overloads are nessesary as no respective parser exists at the 
+    // moment. Remove me when possible.
+    [&](const none&) {
+      return std::make_pair(parse_state::type_not_parsebale,
+                              default_value());
+    },
+    [&](const set&) {
+      return std::make_pair(parse_state::type_not_parsebale,
+                              default_value());
+    },
+    [&](const table&) {
+      return std::make_pair(parse_state::type_not_parsebale,
+                              default_value());
+    },
+    [&](const vector&) {
+      return std::make_pair(parse_state::type_not_parsebale,
+                              default_value());
+    }), default_value_);
+  return result; 
+}
+
+const std::string&
+option_declaration_set::option_declaration::long_name() const {
+  return long_name_;
+}
+
+const std::vector<char>&
+option_declaration_set::option_declaration::short_names() const {
+  return short_names_;
+}
+
+const std::string&
+option_declaration_set::option_declaration::description() const {
+  return description_;
+}
+
+bool option_declaration_set::option_declaration::has_argument() const {
+  return has_argument_;
+}
+
+const data& option_declaration_set::option_declaration::default_value() const {
+  return default_value_;
+}
+
+option_declaration_set::option_declaration_set() {
+  add("help,h?", "print this text", false);
+}
+
+optional<const option_declaration_set::option_declaration&>
+option_declaration_set::find(std::string_view name) const {
+  // TODO: Remove explicit conversion to a string
+  if (auto it = long_opts_.find(std::string{name}); it != long_opts_.end())
+    return *it->second;
+  return {};
+}
+
+std::string option_declaration_set::usage() const {
+  //<--- argument ---> <---- desciption ---->
+  //-w [--write] arg  : path to write events to
+  auto build_argument = [](const option_declaration& x) {
+    std::stringstream arg;
+    auto& shorts = x.short_names();
+    arg << "  "; 
+    if (!shorts.empty()) {
+      arg << "-" <<shorts[0];
+      arg << " [";
+    }
+    if (shorts.size() > 1)
+      for (auto i = 1u; i < shorts.size(); ++i)
+        arg << "-" << shorts[i] << ",";
+    arg << "--" << x.long_name();
+    if (!shorts.empty())
+      arg << "]";
+    if (x.has_argument())
+      arg << " arg";
+    return arg.str();
+  };
+  // calculate the max size the argument column
+  std::vector<size_t> arg_sizes;
+  std::transform(
+    long_opts_.begin(), long_opts_.end(), std::back_inserter(arg_sizes),
+    [&](const auto& x) { return build_argument(*x.second).size(); });
```

Comment:
It's good to use algorithms from the STL. But this has two issues:
1) You are paying for the vector `arg_size`, but all you need is the maximum element.
2) You are throwing away all but the size of `build_argument` only to re-compute it in line 140.

It seems more straightforward to me to have a vector of the arguments instead. Then you can compute the maximum size from that vector and you don't need to compute each argument twice.